File: FrameworkNegocio/LocalizacionesMod/LocalizacionesDN/EmailDN.vb
Imports Framework.DatosNegocio

<Serializable()> _
Public Class EmailDN
    Inherits EntidadDN
    Implements IDatoContactoDN

#Region "Atributos"
    Protected mValorMail As String
    Protected mComentario As String
#End Region

#Region "Propiedades"

    Public Property tipo() As String Implements IDatoContactoDN.Tipo
        Get
            Return Me.GetType.ToString()
        End Get
        Set(ByVal value As String)
            Throw New NotImplementedException()
        End Set
    End Property

    Public Property Valor() As String Implements IDatoContactoDN.Valor
        Get
            Return mValorMail
        End Get
        Set(ByVal value As String)
            CambiarValorVal(Of String)(value, mValorMail)
        End Set
    End Property

    Public Property Comentario() As String Implements IDatoContactoDN.Comentario
        Get
            Return mComentario
        End Get
        Set(ByVal value As String)
            CambiarValorVal(Of String)(value, mComentario)
        End Set
    End Property

#End Region

#Region "Métodos de validación"

    Private Function ValidarValor(ByRef mensaje As String, ByVal valorMail As String) As Boolean
        If Not IsMail(valorMail) Then
            mensaje = "Correo no válido"
            Return False
        End If

        Return True

    End Function

#End Region

#Region "Métodos"

    Public Shared Function IsMail(ByVal address As String) As Boolean
        Dim strRegex As String

        If String.IsNullOrEmpty(address) Then
            Return False
        End If

        strRegex = "^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}" & _
                  "\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\" & _
                  ".)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$"

        Dim re As New System.Text.RegularExpressions.Regex(strRegex)

        If re.IsMatch(address) Then
            Return True
        Else
            Return False
        End If

    End Function

    Public Overrides Function EstadoIntegridad(ByRef pMensaje As String) As Framework.DatosNegocio.EstadoIntegridadDN
        If Not ValidarValor(pMensaje, mValorMail) Then
            Return EstadoIntegridadDN.Inconsistente
        End If

        mToSt = Me.ToString()

        Return MyBase.EstadoIntegridad(pMensaje)
    End Function

#End Region


    
End Class

<Serializable()> _
Public Class ColEmailDN
    Inherits ArrayListValidable(Of EmailDN)

End Class

File: Framework/GestorMensajeria/GestorMails/DN/EmailDN.cs
using System;
using System.Collections.Generic;
using System.Text;

using System.Text.RegularExpressions;

using Framework.DatosNegocio;
using Framework.Mensajeria.GestorMensajeriaDN;

namespace Framework.Mensajeria.GestorMails.DN
{
    [Serializable]
    public class EmailDN : EntidadTipoDN
    {
        #region Atributos

        #endregion

        #region Constructores

        public EmailDN() : base() { }

        public EmailDN(string valorMail) : base(valorMail)
        {
            string mensaje;
            if (!ValidarValor(valorMail, out mensaje))
                throw new ApplicationExceptionDN(mensaje);

            this.modificarEstado = EstadoDatosDN.SinModificar;
        }

        #endregion

        #region Propiedades

        public string ValorMail
        {
            get { return mNombre; }
        }

        #endregion

        #region Métodos Validaciones

        public bool ValidarValor(string valor, out string mensaje)
        {
            mensaje = "";
            if (!IsEmail(valor))
            {
                mensaje = "El mail no es válido";
                return false;
            }
            else
                return true;
        }

        #endregion

        #region Métodos


        public override EstadoIntegridadDN EstadoIntegridad(ref string pMensaje)
        {
            if (!ValidarValor(mNombre, out pMensaje))
                return EstadoIntegridadDN.Inconsistente;
            else
                return EstadoIntegridadDN.Consistente;
        }

        public static bool IsEmail(string address)
        {
            address = address == null ? "" : address;
            string strRegex = @"^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}" +
                  @"\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\" +
                  @".)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$";
            Regex re = new Regex(strRegex);
            if (re.IsMatch(address))
                return true;
            else
                return false;
        }       

        #endregion

    }


    [Serializable]
    public class ColEmailDN : ArrayListValidable<EmailDN>
    {



    }

}
